### PR TITLE
check tasklimit against open tasks

### PR DESCRIPTION
**Pull request by BlueTeck**
possible fix for #4291 might not be the best way

### CODE REVIEW

File: app/Formatter/BoardSwimlaneFormatter.php [hunk 1]
```diff
@@ -101,6 +101,7 @@ public function format()
             foreach ($swimlane['columns'] as $columnIndex => &$column) {
                 $column['column_nb_tasks'] = $this->swimlanes[0]['columns'][$columnIndex]['column_nb_tasks'];
                 $column['column_nb_score'] = $this->swimlanes[0]['columns'][$columnIndex]['column_score'];
+                $column['column_nb_open_tasks'] = $this->taskFinderModel->countByColumnId($column['project_id'], $column['id'] );
```
Comment (fguillot) [inline app/Formatter/BoardSwimlaneFormatter.php]: That might generate too many SQL queries for large board. Need to find another way of fetching the number of open tasks.

Comment (BlueTeck) [reply app/Formatter/BoardSwimlaneFormatter.php]: @fguillot I've changed the implementation, it should be a single select now.

**Comment by BlueTeck** [comment]
@fguillot I've changed the implementation, it should be a single select now.

File: app/Formatter/BoardSwimlaneFormatter.php [hunk 2]
```diff
@@ -104,19 +105,45 @@ public function format()
...
 
     /**
-     * Calculate stats for each column acrosss all swimlanes
+     * Add number of open tasks to eeach column, ignoring the current filter
+     *
+     * @access protected
+     */
+    protected function addNumberOfOpenTasksInColumn()
+    {
+        $column_open_task = $this->db->table(TaskModel::TABLE)
```
Comment (fguillot) [inline app/Formatter/BoardSwimlaneFormatter.php]: This is better but we should avoid doing SQL queries in a Formatter class. This class is about formatting the results from the database.

That would be nice if you could find a way to move this method to the models.

Comment (BlueTeck) [reply app/Formatter/BoardSwimlaneFormatter.php]: @fguillot I've changed it again. It now uses an existing method which returns to taskcount.

**Comment by BlueTeck** [comment]
@fguillot I've changed it again. It now uses an existing method which returns to taskcount.

**Comment by BlueTeck** [comment]
@fguillot whats your opinion on this?

File: app/Formatter/BoardSwimlaneFormatter.php [hunk 3]
```diff
@@ -104,19 +104,39 @@ public function format()
...
 
     /**
-     * Calculate stats for each column acrosss all swimlanes
+     * Add number of open tasks to each column, ignoring the current filter
+     *
+     * @access protected
+     */
+    protected function addNumberOfOpenTasksInColumn()
+    {
+        $columns = $this->columnModel->getAllWithTaskCount($this->swimlanes[0]['columns'][0]['project_id']);
```
Comment (fguillot) [inline app/Formatter/BoardSwimlaneFormatter.php]: If you could move this call in the function `format()`..

File: app/Formatter/BoardSwimlaneFormatter.php [hunk 4]
```diff
@@ -104,19 +104,39 @@ public function format()
...
+
+        if (!empty($columns)) {
+            foreach ($this->swimlanes as &$swimlane) {
+                foreach ($swimlane['columns'] as $columnIndex => &$column) {
+                    $column['column_nb_open_tasks'] = $columns[array_search($column['id'], array_column($columns, 'id'))]['nb_open_tasks'];
```
Comment (fguillot) [inline app/Formatter/BoardSwimlaneFormatter.php]: Is it possible to move this line in the loop above (line 103)? That would avoid doing twice the same loop.

**Comment by BlueTeck** [comment]
@fguillot I refactored it again, I think it even better now. 

Unfortunately the last test failed, but it is not related to this change, just an unlucky timing issue, maybe you could restart the test?

**Comment by fguillot** [comment]
Thanks for addressing my comments.

**Pull request merged**